File: RadioPureWindows/RadioPureWindows/Player/RadioPlayer.cs
// 播放逻辑与 RadioPureShared/RadioPlayer.swift 对齐。

using CommunityToolkit.Mvvm.ComponentModel;
using LibVLCSharp.Shared;
using Microsoft.UI.Dispatching;
using RadioPureWindows.Models;

namespace RadioPureWindows.Player;

public partial class RadioPlayer : ObservableObject, IDisposable
{
    private readonly LibVLC _libVlc;
    private readonly DispatcherQueue _dispatcher;

    private MediaPlayer? _mediaPlayer;
    private Media? _media;
    private string? _resolvedStreamUrl;
    private bool _didTryFallbackForCurrentStation;
    private RadioStation? _pendingStationForFallback;

    [ObservableProperty]
    private bool _isPlaying;

    [ObservableProperty]
    private bool _isError;

    [ObservableProperty]
    private float _volume = 0.7f;

    [ObservableProperty]
    private RadioStation? _currentStation;

    public RadioPlayer(DispatcherQueue dispatcher)
    {
        _dispatcher = dispatcher;
        _libVlc = new LibVLC("--no-video");
    }

    partial void OnVolumeChanged(float value)
    {
        if (_mediaPlayer is not null)
        {
            _mediaPlayer.Volume = (int)(Math.Clamp(value, 0f, 1f) * 100);
        }
    }

    public void Play(RadioStation station)
    {
        if (CurrentStation?.Id == station.Id)
        {
            if (IsPlaying)
            {
                Pause();
            }
            else
            {
                Resume();
            }

            return;
        }

        StopAndRelease(resettingFallback: true);
        CurrentStation = station;
        IsError = false;

        var primary = StreamingUrl(station, primary: true);
        if (primary is null)
        {
            return;
        }

        _resolvedStreamUrl = primary;
        AttachPlayer(station, primary);
    }

    public void Pause()
    {
        _mediaPlayer?.Pause();
        IsPlaying = false;
    }

    public void Resume()
    {
        _mediaPlayer?.Play();
        IsPlaying = true;
        IsError = false;
    }

    public void TogglePlayPause()
    {
        if (IsPlaying)
        {
            Pause();
        }
        else
        {
            Resume();
        }
    }

    private static string? StreamingUrl(RadioStation station, bool primary)
    {
        return primary ? station.Url : station.FallbackUrl;
    }

    private void AttachPlayer(RadioStation station, string streamUrl)
    {
        ReleaseMediaPlayer();

        _pendingStationForFallback = station;
        _mediaPlayer = new MediaPlayer(_libVlc);
        _mediaPlayer.Volume = (int)(Volume * 100);
        _mediaPlayer.EncounteredError += OnEncounteredError;
        _mediaPlayer.EndReached += OnEndReached;

        _media = new Media(_libVlc, streamUrl, FromType.FromLocation);
        _mediaPlayer.Media = _media;
        _mediaPlayer.Play();

        IsPlaying = true;
        IsError = false;
    }

    private void OnEncounteredError(object? sender, EventArgs e)
    {
        DispatchHandlePlaybackFailure();
    }

    private void OnEndReached(object? sender, EventArgs e)
    {
        // 直播流异常结束时也尝试 fallback。
        if (IsPlaying)
        {
            DispatchHandlePlaybackFailure();
        }
    }

    private void DispatchHandlePlaybackFailure()
    {
        _dispatcher.TryEnqueue(() =>
        {
            var station = _pendingStationForFallback ?? CurrentStation;
            if (station is not null)
            {
                HandlePlaybackFailure(station);
            }
        });
    }

    private void HandlePlaybackFailure(RadioStation station)
    {
        var fallback = station.FallbackUrl;
        if (fallback is not null &&
            !_didTryFallbackForCurrentStation &&
            _resolvedStreamUrl != fallback)
        {
            _didTryFallbackForCurrentStation = true;
            IsError = false;
            _resolvedStreamUrl = fallback;
            AttachPlayer(station, fallback);
            return;
        }

        IsPlaying = false;
        IsError = true;
        _pendingStationForFallback = null;
    }

    private void StopAndRelease(bool resettingFallback = false)
    {
        ReleaseMediaPlayer();
        IsPlaying = false;
        IsError = false;
        _resolvedStreamUrl = null;

        if (resettingFallback)
        {
            _didTryFallbackForCurrentStation = false;
            _pendingStationForFallback = null;
        }
    }

    private void ReleaseMediaPlayer()
    {
        if (_mediaPlayer is not null)
        {
            _mediaPlayer.EncounteredError -= OnEncounteredError;
            _mediaPlayer.EndReached -= OnEndReached;
            _mediaPlayer.Stop();
            _mediaPlayer.Dispose();
            _mediaPlayer = null;
        }

        _media?.Dispose();
        _media = null;
    }

    public void Dispose()
    {
        StopAndRelease(resettingFallback: true);
        _libVlc.Dispose();
        GC.SuppressFinalize(this);
    }
}

File: RadioPureAndroid/app/src/main/kotlin/com/radiopure/app/radiopure/player/RadioPlayer.kt
package com.radiopure.app.radiopure.player

import android.net.Uri
import androidx.media3.common.MediaItem
import androidx.media3.common.MediaMetadata
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer
import com.radiopure.app.radiopure.model.RadioStation
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

/**
 * 播放逻辑与 [RadioPureShared/RadioPlayer.swift] 对齐。
 */
class RadioPlayer(
    private val exoPlayer: ExoPlayer,
) {
    private val _state = MutableStateFlow(RadioPlayerState())
    val state: StateFlow<RadioPlayerState> = _state.asStateFlow()

    private var resolvedStreamURLString: String? = null
    private var didTryFallbackForCurrentStation = false
    private var pendingStationForFallback: RadioStation? = null

    private val playerListener = object : Player.Listener {
        override fun onIsPlayingChanged(isPlaying: Boolean) {
            _state.update { it.copy(isPlaying = isPlaying && exoPlayer.playbackState != Player.STATE_IDLE) }
        }

        override fun onPlayerError(error: androidx.media3.common.PlaybackException) {
            val station = pendingStationForFallback ?: _state.value.currentStation ?: return
            handlePlaybackFailure(station)
        }
    }

    init {
        exoPlayer.addListener(playerListener)
        exoPlayer.volume = _state.value.volume
    }

    fun play(station: RadioStation) {
        val current = _state.value
        if (current.currentStation?.id == station.id) {
            if (current.isPlaying) pause() else resume()
            return
        }

        stopAndRelease(resettingFallback = true)
        _state.update {
            it.copy(
                currentStation = station,
                isError = false,
                isPlaying = false,
            )
        }
        val primary = streamingURL(station, primary = true) ?: return
        resolvedStreamURLString = primary
        attachPlayer(station, Uri.parse(primary))
    }

    fun pause() {
        exoPlayer.pause()
        _state.update { it.copy(isPlaying = false) }
    }

    fun resume() {
        exoPlayer.play()
        _state.update { it.copy(isPlaying = true, isError = false) }
    }

    fun togglePlayPause() {
        if (_state.value.isPlaying) pause() else resume()
    }

    fun setVolume(volume: Float) {
        val clamped = volume.coerceIn(0f, 1f)
        exoPlayer.volume = clamped
        _state.update { it.copy(volume = clamped) }
    }

    fun detach() {
        exoPlayer.removeListener(playerListener)
        stopAndRelease(resettingFallback = true)
    }

    private fun streamingURL(station: RadioStation, primary: Boolean): String? {
        return if (primary) station.url else station.fallbackURL
    }

    private fun attachPlayer(station: RadioStation, streamUri: Uri) {
        pendingStationForFallback = station
        val metadata = MediaMetadata.Builder()
            .setTitle(station.name)
            .setArtist("RadioPure")
            .build()
        val mediaItem = MediaItem.Builder()
            .setUri(streamUri)
            .setMediaMetadata(metadata)
            .build()
        exoPlayer.setMediaItem(mediaItem)
        exoPlayer.prepare()
        exoPlayer.play()
        _state.update { it.copy(isPlaying = true, isError = false) }
    }

    private fun handlePlaybackFailure(station: RadioStation) {
        val fallback = station.fallbackURL
        if (
            fallback != null &&
            !didTryFallbackForCurrentStation &&
            resolvedStreamURLString != fallback
        ) {
            didTryFallbackForCurrentStation = true
            _state.update { it.copy(isError = false) }
            resolvedStreamURLString = fallback
            exoPlayer.stop()
            exoPlayer.clearMediaItems()
            attachPlayer(station, Uri.parse(fallback))
            return
        }

        _state.update { it.copy(isPlaying = false, isError = true) }
        pendingStationForFallback = null
    }

    private fun stopAndRelease(resettingFallback: Boolean) {
        exoPlayer.stop()
        exoPlayer.clearMediaItems()
        _state.update { it.copy(isPlaying = false, isError = false) }
        resolvedStreamURLString = null
        if (resettingFallback) {
            didTryFallbackForCurrentStation = false
            pendingStationForFallback = null
        }
    }
}
